### PR TITLE
Implement fees in outbound router since it is not supported

### PR DESCRIPTION
**Pull request by claravanstaden**
Remove support for target chain fees in the XCM message, since they are not handled in the target chain at all. In any case, target chain fees will be added in: https://github.com/Snowfork/snowbridge/pull/924

Resolves: [SNO-645](https://linear.app/snowfork/issue/SNO-645)
Cumulus companion: https://github.com/Snowfork/cumulus/pull/62

### CODE REVIEW

File: parachain/primitives/router/src/outbound/mod.rs [hunk 1]
```diff
@@ -103,16 +103,11 @@ where
 		})?;
 
 		let mut converter = XcmConverter::new(&message, &gateway_network, &gateway_address);
-		let (agent_execute_command, max_target_fee) = converter.convert().map_err(|err|{
+		let (agent_execute_command, _max_target_fee) = converter.convert().map_err(|err|{
```
Comment (claravanstaden) [inline parachain/primitives/router/src/outbound/mod.rs:106]: I guess `_max_target_fee` can be removed in its entirety here.

File: parachain/primitives/router/src/outbound/mod.rs [hunk 2]
```diff
@@ -216,12 +209,6 @@ impl<'a, Call> XcmConverter<'a, Call> {
 	fn fee_info(&mut self) -> Result<Option<&'a MultiAsset>, XcmConverterError> {
 		use XcmConverterError::*;
 		let execution_fee = match self.next()? {
-			WithdrawAsset(fee_asset) => match self.next()? {
-				BuyExecution { fees: execution_fee, weight_limit: Unlimited }
-					if fee_asset.len() == 1 && fee_asset.contains(execution_fee) =>
-					Some(execution_fee),
-				_ => return Err(BuyExecutionExpected),
-			},
 			UnpaidExecution { check_origin: None, weight_limit: Unlimited } => None,
```
[the remark this answers is not in the document]
Comment (yrong) [reply parachain/primitives/router/src/outbound/mod.rs:212]: https://github.com/paritytech/polkadot-sdk/blob/2f49252bcdb450c36f83a6dacdececc7b5d8e32c/polkadot/xcm/xcm-executor/src/lib.rs#L853

https://github.com/paritytech/polkadot-sdk/blob/2f49252bcdb450c36f83a6dacdececc7b5d8e32c/polkadot/xcm/xcm-executor/src/lib.rs#L956-L957

https://github.com/paritytech/polkadot-sdk/blob/2f49252bcdb450c36f83a6dacdececc7b5d8e32c/cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/xcm_config.rs#L255

So it requires a properly configured `FeeManager` as first step.

[the remark this answers is not in the document]
Comment (alistair-singh) [reply parachain/primitives/router/src/outbound/mod.rs]: I am not sure if removing this code resolves the audit issue. The audit report suggests handling both scenarios. This PR makes no change to the observable behavior of this code. Both the old code and the new code will return `SendError::Unroutable` when target fees are specified. The only change is what error gets logged out to the log file. The new code will log TargetFeesExpected as the error. The old code will log TargetFeesNotSupported as the error, which I think is more self documenting.

So maybe this should be solved in the way Ron suggested.

[the remark this answers is not in the document]
Comment (claravanstaden) [reply parachain/primitives/router/src/outbound/mod.rs]: Thanks for the feedback.

@yrong I understood these fees as fees on the target chain, right? So the fees also need to be paid on the Ethereum side?

[the remark this answers is not in the document]
Comment (claravanstaden) [reply parachain/primitives/router/src/outbound/mod.rs]: @alistair-singh Ok sure! I thought we would just remove the fee XCM instructions and only allow UnpaidExecution for now (I know it doesn't really make a change to the functionality, but neither does some of the other refactor issues), but if you think it's better to implement the fees now I will get on it.

[the remark this answers is not in the document]
Comment (yrong) [reply parachain/primitives/router/src/outbound/mod.rs]: In the end all incurred costs(including gas cost on the Ethereum side) should be covered by the original sender(i.e. a signed origin like Alice or Bob in assetHub). 

Though we have some `fee_info` logic prepared to extract fee from the original xcm, we don't support it as you can see the check [here](https://github.com/Snowfork/snowbridge/blob/0cd56457f6e96f0dc6111027a7e6caccea60217d/parachain/primitives/router/src/outbound/mod.rs#L111-L114) actually only unpaid is allowed.

For fully implementing the fee charge mechanism it requires several other xcm components as I mentioned above which still missing(considerable change required both in assetHub and bridgeHub).

I think the current implementation should be fine for demo purposes or the launch on Rococo testnet.  But definitely not for production usage as you mentioned before the kusama or polkadot.

I don't know if we would implement it right now or later, leave to the team for the final decision.

[the remark this answers is not in the document]
Comment (vgeddes) [reply parachain/primitives/router/src/outbound/mod.rs]: I mostly agree with Ron. In general there are some missing pieces in BridgeHub, and we are on a very old version of Polkadot/Substrate/Cumulus/. So it is hard to make sense of where we need to be.

The big difference is that in the newer versions, `AssetHub` will not be configured with a `SovereignRemoteExporter` or a `UnpaidRemoteExporter`. Instead AssetHub will directly send an XCM message to BridgeHub containing an `ExportMessage` instruction.

So I think we should consider this PR _BLOCKED_ for the time being.

[the remark this answers is not in the document]
Comment (yrong) [reply parachain/primitives/router/src/outbound/mod.rs]: Seems they're just trying to reuse the previous `reserve_transfer_assets` without introducing a new pallet for this?

**Comment by claravanstaden** [comment]
Closing this PR since https://github.com/Snowfork/snowbridge/pull/940#pullrequestreview-1644409319 covers this.

**Pull request closed without merging**